### PR TITLE
Feat/ Redis HA fix

### PR DESCRIPTION
**Pull request by alithethird**
Includes the [Redis HA fix](https://github.com/canonical/redis-k8s-operator/pull/96). This fixes 2 different bugs: 
1- https://bugs.launchpad.net/juju/+bug/2063087
2- https://github.com/canonical/discourse-k8s-operator/issues/268

### Overview

Started using app databag for Redis hostname. Add tests to test out Redis in HA mode. 

### Rationale

Previously was not able to use Redis in HA mode

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Manging Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -390,16 +390,17 @@ def _get_redis_relation_data(self) -> typing.Tuple[str, int]:
...
-        relation_data = self.redis.relation_data
-        if not relation_data:
+        relation = self.model.get_relation(self.redis.relation_name)
+        if not relation:
             raise MissingRedisRelationDataError("No redis relation data")
 
         try:
-            redis_hostname = str(relation_data["hostname"])
-            redis_port = int(relation_data["port"])
+            relation_data = relation.data[self.model.relations["redis"][0].app]
```
Comment (javierdelapuente) [inline src/charm.py:398]: Maybe something like this:
`relation_data_app = relation.data[relation.app]`.

**Review by javierdelapuente — COMMENTED** [review]
One question, what are now the restrictions to integrate with Redis? A specific version?

Should we fallback to the previous functionality if "leader-host" is not there? Otherwise, should we tell the user that this charm requires a Redis charm revision greater than X?

**Comment by alithethird** [comment]
> One question, what are now the restrictions to integrate with Redis? A specific version?
> 
> Should we fallback to the previous functionality if "leader-host" is not there? Otherwise, should we tell the user that this charm requires a Redis charm revision greater than X?

Ok, that is a great question. I did not think of that. I think I should ask @gregory-schiano about this. 

I think the fallback to unit data if `leader-host` doesn't exist is a great idea. What do you think?

File: src/charm.py [hunk 2]
```diff
@@ -390,16 +390,23 @@ def _get_redis_relation_data(self) -> typing.Tuple[str, int]:
...
-            redis_hostname = str(relation_data["hostname"])
-            redis_port = int(relation_data["port"])
+            if not redis_hostname:
+                redis_hostname = str(self.redis.relation_data["hostname"])
+            redis_port = int(self.redis.relation_data["port"])
         except (KeyError, ValueError) as exc:
             raise MissingRedisRelationDataError(
-                "Wrong hostname or port in Redis relation"
+                "Wrong hostname or port in Redis App data"
             ) from exc
```
Comment (gregory-schiano) [inline src/charm.py:410]: The whole structure of the code bothers me a bit, can't we do something like:
```python
try:
   redis_port = int(relation_data["port"])
   redis_hostname = str(relation_app_data["leader-host"] if relation_app_data.get("leader-host") else relation_app_data["hostname"])
except (KeyError, ValueError) as exc:
   raise MissingRedisRelationDataError("Either 'leader-host' or 'hostname' and 'ports' are mandatory") from exc
```

Or something around the lines

Comment (alithethird) [reply src/charm.py:410]: Sure can do! I will also do the same in Indico Charm.

File: src/charm.py [hunk 3]
```diff
@@ -390,16 +390,22 @@ def _get_redis_relation_data(self) -> typing.Tuple[str, int]:
...
+        relation = self.model.get_relation(self.redis.relation_name)
+        if not relation:
             raise MissingRedisRelationDataError("No redis relation data")
+        relation_app_data = relation.data[relation.app]
+        relation_unit_data = self.redis.relation_data
 
         try:
-            redis_hostname = str(relation_data["hostname"])
-            redis_port = int(relation_data["port"])
+            redis_hostname = str(
```
Comment (javierdelapuente) [inline src/charm.py:400]: As an alternative:

`redis_hostname = str(relation_app_data.get("leader-host", relation_unit_data["hostname"]))`

All the catching of the exceptions is a bit strange, but can be refactored in the future as it is not what this change did.

**Review by javierdelapuente — APPROVED** [review]

File: src/charm.py (hunk 3 above)
Comment (alithethird) [reply src/charm.py:400]: Thats even cleaner! Thank you :D. I will remember this if I do a refactor pr in the future.

**Review by amandahla — APPROVED** [review]
LGTM

**Pull request merged**